### DISCUSSION TITLE
TimePicker alternative - click position on clock image

### DISCUSSION THREAD

**Discussion by bsdRuf** [AvaloniaUI/Avalonia#17138] [General]
Hello there, 

I wasnt able to find much alternatives to the default TimePicker.

Thats why I made a flyout containing a image of a clock. Where I calculate the angle from middle to clicked position. The image has to be a square to get the correct time.

Hour wheel
![image](https://github.com/user-attachments/assets/aa28ce29-fabb-48c1-b3ee-e9b792eff114)
Minute wheel
![image](https://github.com/user-attachments/assets/cd0aa5cc-d7c0-4069-9388-be01ee0492cc)

If you optimize anything, I would be glad if you let me know
Repo: [GitHub](https://github.com/bsdRuf/TimePickerExample)

Thanks for the great work
Avalonia team and devs

**Comment by robloo** [comment]
This is great! Would you be willing to polish it up and merge it into Avalonia itself if the core team allows it? I also think it should be renamed to `ClockTimePicker` which aligns with `CalendarDatePicker` and also doesn't conflict with existing control names.

Looking at the source code it would need to be re-architected though. It should not use an image and needs to be drawn graphics both for performance and size. It's also not structured as a templated control so would need to be re-thought from the ground up. However, the concept is good and the visual design is a great start.

This is a fundamental missing piece in Xaml frameworks I've talked about before. The full summary is here: https://github.com/microsoft/microsoft-ui-xaml/issues/3605 over on the WinUI repository.

Basically there are 6 controls that are needed to cover all cases and we only have three. I think most people, like you, intuitively know we are missing the 4th. It's also in used in Android quite a bit. It would be great to start filling in gaps here.

DatePicker : A control for fast numerical input of dates
TimePicker : A control for fast numerical input of times
CalendarDatePicker : A control for visual (skeuomorphic) selection of a date on a calendar
**ClockTimePicker : A control for visual (skeuomorphic) selection of a time on a clock**

The other two missing controls combine both Date/Time selection together. Those would have to be a different conversation but basically a DateTimePicker would allow selection of BOTH a date and time in the same control. There would also be a variant of the DateTimePicker that has both a calendar and a clock -- probably in separate tabs like the ColorPicker. Some OSes have this type of control already but it is more rare.

![image](https://github.com/user-attachments/assets/27942bc9-5806-43ad-b616-fcb0bc435603)
